### PR TITLE
Patch ldflags and libs in GCC_compile under windows

### LINKED ISSUE #944: Aesara fails to set the correct default blas ldflags to link with MKL on windows
## Description of your problem or feature request

MKL libraries on windows are supplied as dynamic link libraries (e.g. `mkl_core.1.dll`) and don't have the lib prefix. When aesara tries to set the default blas ldflags [here](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L2635) it does a few things:

1. It tries to import `mkl`, which should be available if `mkl-services` was installed via conda or mamba.
2. Sets up a gcc styled library directory flag `-L` [see here](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L2763)
3. It sets the default libraries to try to link: `-lmkl_core -lmkl_rt -lmkl_intel_thread`
4. It then tries to compile and run a small [test script that links to mkl](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L1961-L1993)

**Please provide the full traceback of any errors.**
The test script fails to compile and run [here](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L1985-L1987). You can run it manually using this snippet:

```python
import textwrap
from aesara.link.c.cmodule import std_lib_dirs, GCC_compiler
import os
import sys


flags = [
    '-L"C:\\Users\\chulo\\Anaconda3\\Library\\bin"',
    '-lmkl_core',
    '-lmkl_intel_thread',
    '-lmkl_rt',
]


test_code = textwrap.dedent(
    """\
    extern "C" double ddot_(int*, double*, int*, double*, int*);
    int main(int argc, char** argv)
    {
        int Nx = 5;
        int Sx = 1;
        double x[5] = {0, 1, 2, 3, 4};
        double r = ddot_(&Nx, x, &Sx, x, &Sx);
        if ((r - 30.) > 1e-6 || (r - 30.) < -1e-6)
        {
            return -1;
        }
        return 0;
    }
    """
)
cflags = list(flags)
# to support path that includes spaces, we need to wrap it with double quotes on Windows
path_wrapper = '"' if os.name == "nt" else ""
cflags.extend([f"-L{path_wrapper}{d}{path_wrapper}" for d in std_lib_dirs()])

res = GCC_compiler.try_compile_tmp(
    test_code, tmp_prefix="try_blas_", flags=cflags, try_run=True, output=True
)
```

This results in
```python
>>> print(res[-1].decode("utf-8"))

C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lmkl_core: No such file or directory
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lmkl_intel_thread: No such file or directory
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lmkl_rt: No such file or directory
collect2.exe: error: ld returned 1 exit status
```

**Please provide any additional information below.**

Since the mkl .dll libraries don't have the lib prefix, we should provide their absolute paths to be able to link them when compiling. This means that setting the flags to:

```python
import sys

flags = [
    os.path.join(sys.prefix, "Library", "bin", "mkl_core.1.dll"),
    os.path.join(sys.prefix, "Library", "bin", "mkl_intel_thread.1.dll"),
    os.path.join(sys.prefix, "Library", "bin", "mkl_rt.1.dll"),
]
```

makes the compilation succeed.

## Versions and main components

* Aesara version: 2.6.2
* Aesara config (`python -c "import aesara; print(aesara.config)"`)
* Python version: 3.9.7
* Operating system: Windows 10
* How did you install Aesara: (conda/pip) conda

### PR DESCRIPTION
**Pull request by lucianopaz**
Closes #944

On windows systems, the dll or lib libraries don't follow gcc's library naming convention, which means that their file name doesn't have the lib prefix (e.g. libmkl_core.dll). For this reason, gcc's library search fails to find these dynamic link libraries or static libs. An undesired consequence was that aesara (and old theano) was not able to set the correct blas__ldflags to link with mkl, even if the library had been correctly installed.

This PR introduces two GCC_compiler static methods that patch the libs or ldflags that are used during compilation. The static methods specifically target dll and lib files. They search for the library files in the provided library directories, and if a dll or lib is found, the ldflag is set to the absolute path that points to the dll or lib file. As a sidenote, dlls are preferred over libs.

As a consequence, the normal unix-like ldflags will be patched to work in windows as they do in unix. Thanks to this, aesara is now able to set the correct ldflags in windows to link to mkl installations.

### CODE REVIEW

**Comment by lucianopaz** [comment]
Does anyone know how to fix the unreachable code error from mypy? That code only runs on windows, and mypy doesn’t complain when I run it on windows.

**Comment by lucianopaz** [comment]
It looks like mypy is failing because of [this bug](https://github.com/python/mypy/issues/12545)

**Comment by brandonwillard** [comment]
> It looks like mypy is failing because of [this bug](https://github.com/python/mypy/issues/12545)

It looks more like this one: https://github.com/python/mypy/issues/5678.

**Comment by lucianopaz** [comment]
The failing test (`tests/sparse/sandbox/test_sp.py::TestSP::test_maxpool`) passes locally. Maybe the test is flaky?

**Comment by brandonwillard** [comment]
> The failing test (`tests/sparse/sandbox/test_sp.py::TestSP::test_maxpool`) passes locally. Maybe the test is flaky?

Yeah, that looks like a bad, unseeded test.

File: aesara/link/c/cmodule.py [hunk 1]
```diff
@@ -2395,9 +2402,59 @@ def try_flags(
...
+        def sort_key(lib):  # type: ignore
+            name, *numbers, extension = lib.split(".")
+            return (extension == "dll", tuple(map(int, numbers)))
+
+        patched_lib_ldflags = []
+        for lib in libs:
+            ldflag = f"-l{lib}"
+            for lib_dir in lib_dirs:
+                if lib_dir.startswith('"') and lib_dir.endswith('"'):
+                    lib_dir = lib_dir[1:-1]
```
Comment (brandonwillard) [inline aesara/link/c/cmodule.py:2444]: Looks like we need coverage for this scenario.  Also, isn't this equivalent to `lib_dir = lib_dir.strip('"')`?  If it is, then you can execute that statement unconditionally.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by hectormz** [comment]
Looking forward to this PR! Was just scratching my head about this on Windows. Thanks @lucianopaz

**Comment by lucianopaz** [comment]
Codecov seems to be linking the old commit instead of the [new one](https://codecov.io/github/aesara-devs/aesara/commit/0dc06efe59cf0455a5b9a5f7d14e587e3cde714e)

**Pull request merged**